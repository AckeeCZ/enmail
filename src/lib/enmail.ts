export interface Message {
    from?: string;
    to?: string | string[];
    content?: string;
    subject?: string;
}

export type SendMessage<AdapterOptions = any, SendOptions = any> =
    (message: Message, adapter: AdapterSend | Adapter<AdapterOptions, SendOptions>, sendOptions?: SendOptions) => Promise<unknown>;
export type AdapterSend<SendOptions = any> = (message: Message, options?: SendOptions) => Promise<unknown>;

export interface Adapter<AdapterOptions = any, SendOptions = any> {
    options: AdapterOptions;
    mailService: any;
    sender: (sendOptions?: SendOptions) => AdapterSend<SendOptions>;
}
