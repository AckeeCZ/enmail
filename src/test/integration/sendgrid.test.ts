import enmail from '../../index';
import { mockApis } from '../utils/testMock';

describe('Sendgrid adapter', () => {
    const getSendgridAdapter = () => enmail.createAdapter.sendgrid(options);
    const options = { apiKey: 'SG.apiKey' };
    const adapter = getSendgridAdapter();
    const sendOptions = { mailSettings: { sandboxMode: { enable: true } } };
    const body = {
        to: 'test@example.com',
        from: 'test2@example.com',
        content: 'Hello world',
        subject: 'Hello',
    };
    beforeAll(() => mockApis());
    it('Sendgrid adapter can be created', () => {
        expect(typeof adapter.sender).toEqual('function');
        expect(adapter.options).toStrictEqual(options);
        expect(typeof adapter.mailService.send).toEqual('function');
        expect(adapter).toMatchSnapshot();
    });
    it('Sender will be successfully returned', () => {
        const sender = enmail.sender(adapter);
        expect(typeof sender).toEqual('function');
        expect(sender).toEqual(adapter.sender);
    });
    it('An email can be sent via adapter', async () => {
        const [{ statusCode }] = await adapter.sender()(body, sendOptions) as any;
        expect(statusCode).toEqual(200);
    });
    it('An email can be sent via adapter with send options', async () => {
        const [{ statusCode }] = await adapter.sender(sendOptions)(body) as any;
        expect(statusCode).toEqual(200);
    });
    it('An email can be sent via adapter without send options', async () => {
        const [{ statusCode }] = await adapter.sender()(body) as any;
        expect(statusCode).toEqual(200);
    });
    it('An email can be sent via enmail.send', async () => {
        const [{ statusCode }] = await enmail.send(body, adapter) as any;
        expect(statusCode).toEqual(200);
    });
    it('An email can be sent via enmail.send with sendOptions', async () => {
        const [{ statusCode }] = await enmail.send(body, adapter, { myOptions: 'options' }) as any;
        expect(statusCode).toEqual(200);
    });
    it('An email can be sent via sender', async () => {
        const [{ statusCode }] = await enmail.send(body, adapter.sender()) as any;
        expect(statusCode).toEqual(200);
    });
});
