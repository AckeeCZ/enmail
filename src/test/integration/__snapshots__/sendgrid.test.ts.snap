// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sendgrid adapter Sendgrid adapter can be created 1`] = `
Object {
  "mailService": MailService {
    "MailService": [Function],
    "client": Client {
      "auth": "Bearer SG.apiKey",
      "defaultHeaders": Object {
        "Accept": "application/json",
        "Content-Type": "application/json",
        "User-Agent": "sendgrid/7.1.1;nodejs",
      },
      "defaultRequest": Object {
        "baseUrl": "https://api.sendgrid.com/",
        "headers": Object {},
        "maxContentLength": Infinity,
        "method": "GET",
        "url": "",
      },
      "impersonateSubuser": "",
    },
    "secretRules": Array [],
    "substitutionWrappers": Array [
      "{{",
      "}}",
    ],
  },
  "options": Object {
    "apiKey": "SG.apiKey",
  },
  "sender": [Function],
}
`;
